main.go

```diff
@@ -49,17 +50,134 @@ func makeClient() http.Client {
 	return proxyClient
 }
 
+type NatsQueue struct {
+	clusterID string
+	clientID  string
+	natsURL   string
+
+	maxReconnect   int
+	reconnectDelay time.Duration
+	conn           stan.Conn
+	connMutex      *sync.RWMutex
+	quitCh         chan struct{}
+
+	subject        string
+	qgroup         string
+	durable        string
+	ackWait        time.Duration
+	messageHandler func(*stan.Msg)
+	startOption    stan.SubscriptionOption
+	maxInFlight    stan.SubscriptionOption
+	subscription   stan.Subscription
+}
+
+func (q *NatsQueue) init() error {
+	q.connMutex.Lock()
+	defer q.connMutex.Unlock()
+
+	log.Printf("Connecting to: %s\n", q.natsURL)
+
+	sc, err := stan.Connect(
+		q.clusterID,
+		q.clientID,
+		stan.NatsURL(q.natsURL),
+		stan.SetConnectionLostHandler(func(conn stan.Conn, err error) {
+			log.Printf("Disconnected from %s\n", q.natsURL)
+
+			q.reconnect()
+		}),
+	)
+	if err != nil {
+		return fmt.Errorf("Can't connect to %s: %v\n", q.natsURL, err)
+	}
+
+	q.conn = sc
```

Comment:
I wonder if we can move the lock closer to where we get/set the value? The value is only protecting `q.conn` right?

Reply:
Done